### PR TITLE
WIP - Custom readiness queue

### PR DESCRIPTION
**Pull request by carllerche**
WIP, not ready for merging. This PR will be the first step towards #360

### CODE REVIEW

File: src/poll.rs [hunk 1]
```diff
@@ -42,14 +48,133 @@ use std::time::Duration;
...
+
+    // The Token used to register the `Evented` with `Poll`. This can change,
+    // but only by calling `Poll` functions, so there will be no concurrency.
+    token: Token,
+
+    // Both interest and opts can be mutated
+    interest: Cell<EventSet>,
+
+    // Poll opts
+    opts: Cell<PollOpt>,
```
Comment (alexcrichton) [inline src/poll.rs]: I'm personally pretty nervous about the direction this abstraction is headed in. As is it seems like it's incredibly difficult to get right in terms of preventing race conditions and other undefined behavior, and it's barely leveraging Rust types for managing this unsafety. Specifically this structure is conglomeration of:

* Raw unsafe pointers
* non-thread-safe `Cell` types
* threadsafe atomics

This, combined with the `fn inner_mut(&self) -> &mut ReadinessQueueInner` signature below, make this very difficult to determine whether it's actually safe or not.

I think this would benefit quite a bit from extracting some of the unsafe components here to standalone and isolated features with safe interfaces, or very very clear unsafe interfaces with documentation as to what contracts need to be upheld. For example `next_all_notes` and `prev_all_nodes` looks a lot like a doubly linked list? Does the standard library's `LinkedList` not suffice here because atomics are necessary? Some other reason?

Similarly if there's a set of fields that are only accessed by a `Poll` on each node, then this may benefit from a more concrete interface to indicate such. For example I would expect something like:

```rust
struct PollFields { ... }

struct ReadinessNode {
    poll_fields: UnsafeCell<PollFields>,
}

impl ReadinessNode {
    fn poll_fields<'a>(&'a self, _proof: &'a Poll) -> &'a PollFields { /* ... */ }
}
```

Basically something like that where the API itself is very difficult to get wrong if properly used. That should help anyone else working around here or perhaps any future refactorings as well.

File: src/poll.rs [hunk 2]
```diff
@@ -145,3 +270,339 @@ pub fn selector(poll: &Poll) -> &sys::Selector {
...
+        // If the node is currently ready, re-queue?
+        if !event::is_empty(self.readiness()) {
+            self.queue_for_processing(None);
+        }
+
+        Ok(())
+    }
+
+    pub fn readiness(&self) -> EventSet {
+        event::from_usize(self.node().events.load(Ordering::Relaxed))
```
Comment (alexcrichton) [inline src/poll.rs]: I would personally **very strongly recommend** using `SeqCst` everywhere. It's far easier to selectively relax orderings, and profiling can be used to figure out which ones are on the hot paths.

If everything is Relaxed or Acquire/Release it's incredibly difficult to reason about whether the orderings are actually correct. They're all basically related to global decisions, so you have to understand basically everything to reason about any local ordering as well. Thankfully if most orderings are `SeqCst` you can safely ignore most of the program as they'll work "as expected".

File: src/poll.rs [hunk 3]
```diff
@@ -145,3 +270,339 @@ pub fn selector(poll: &Poll) -> &sys::Selector {
...
+        let queue = poll.readiness_queue.clone();
+        let node = queue.new_readiness_node(token, interest, opts);
+
+        Registration {
+            node: node,
+            queue: queue,
+        }
+    }
+
+    pub fn update(&self, poll: &Poll, interest: EventSet, opts: PollOpt) -> io::Result<()> {
```
Comment (alexcrichton) [inline src/poll.rs]: Since this `Registration` is already inherently tied to the original `Poll`, is there a benefit from taking it as an extra parameter here? It'd perhaps be nicer to not have to pass it?

File: src/poll.rs [hunk 4]
```diff
@@ -145,3 +270,339 @@ pub fn selector(poll: &Poll) -> &sys::Selector {
...
+        if !self.queue.identical(&poll.readiness_queue) {
+            return Err(io::Error::new(io::ErrorKind::Other, "nope"));
+        }
+
+        self.node().interest.set(interest);
```
Comment (alexcrichton) [inline src/poll.rs]: This is kinda surprising to me in terms of what's going on. We're modifying a `Cell` which is explicitly documented as only accessible from `Poll`, which sounds like a data race waiting to happen?

This is sort of what I was alluding to above with where the unsafety here may be a little too much to reason about. It's very unclear which operations are unsafe and which aren't because the `unsafe` blocks are so wildly powerful in this module.

**Comment by carllerche** [comment]
Looks like CI is green. I'm hoping to merge this!

File: src/channel.rs [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    (tx, rx)
+}
+
+impl SenderCtl {
+    /// Call to track that a message has been sent
+    pub fn inc(&self) -> io::Result<()> {
+        if 0 == self.inner.pending.fetch_add(1, Ordering::Acquire) {
+            // Toggle readiness to readable
+            if let Some(set_readiness) = self.inner.set_readiness.as_ref() {
+                // TODO: Don't ignore result
```
Comment (alexcrichton) [inline src/channel.rs]: This TODO is handled, right?

File: src/channel.rs [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    let rx = ReceiverCtl {
+        registration: Lazy::new(),
+        inner: inner,
+    };
+
+    (tx, rx)
+}
+
+impl SenderCtl {
+    /// Call to track that a message has been sent
```
Comment (alexcrichton) [inline src/channel.rs]: Maybe worth documenting what can be done in the case of an error? E.g. the readiness is not set, but the increment still happened

File: src/channel.rs [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        match *self {
+            StdSender::Bounded(ref v) => StdSender::Bounded(v.clone()),
+            StdSender::Unbounded(ref v) => StdSender::Unbounded(v.clone()),
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    pub fn try_recv(&self) -> Result<T, mpsc::TryRecvError> {
+        // TODO: When there is no value to recv, unset readiness
```
Comment (alexcrichton) [inline src/channel.rs]: This is done through `dec`, right?

File: src/lazy.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+impl<T> Lazy<T> {
+    pub fn new() -> Lazy<T> {
+        Lazy { val: UnsafeCell::new(None) }
+    }
+
+    pub fn is_some(&self) -> bool {
+        self.as_ref().is_some()
+    }
```
Comment (alexcrichton) [inline src/lazy.rs]: I don't think this is a safe method to expose because concurrent calls to `Lazy::is_some` and `Lazy::set` is a data race

Comment (alexcrichton) [inline src/lazy.rs]: er nvmd, I'm getting confused between these two

Comment (alexcrichton) [inline src/lazy.rs]: er actually, this `Lazy` type seems like it's not used?

Comment (carllerche) [inline src/lazy.rs]: `Lazy` is used here https://github.com/carllerche/mio/blob/custom-readiness-queue/src/channel.rs#L96

File: src/channel.rs (hunk 7 above)
Comment (carllerche) [inline src/channel.rs]: Yes, `dec` will unset readiness. I think I was suggesting something else... I also think the TODO here is out of date. I will remove.

**Comment by rrichardson** [comment]
LGTM

File: src/channel.rs [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            tx: self.tx.clone(),
+            ctl: self.ctl.clone(),
+        }
+    }
+}
+
+impl<T> StdSender<T> {
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        match *self {
+            StdSender::Bounded(ref tx) => tx.send(t).map_err(SendError::from),
```
Comment (hjr3) [inline src/channel.rs]: Shouldn't this be using `try_send`?
